core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java

```diff
@@ -133,29 +200,107 @@ public GeoBoundingBoxQueryBuilder topRight(String geohash) {
+        }
+
+        if (validationException == null) {
+            // all corners are valid after above checks - make sure they are in the right relation
+            if (topLeft.getLat() < bottomRight.getLat()) {
+                validationException = addValidationError("Top part of top-left corner is below bottom part of bottom-right corner: " +
+                        topLeft.getLat() + " vs. " + bottomRight.getLat(), validationException);
+            }
+            if (topLeft.getLon() > bottomRight.getLon()) {
+                validationException = addValidationError("Left part of top-left corner is right of right part of bottom-right corner: " +
+                        topLeft.getLon() + " vs. " + bottomRight.getLon(), validationException);
+            }
         }
 
+        if (Strings.isEmpty(fieldName)) {
+            validationException = addValidationError("FieldName must be non-null and non-empty.", validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    public Query doToQuery(QueryParseContext parseContext) {
+        if (parseContext.mapperService().smartNameFieldType(fieldName, parseContext.getTypes()) == null){
+            return new MatchNoDocsQuery();
+        }
+
+        MappedFieldType fieldType = parseContext.fieldMapper(fieldName);
+        if (fieldType == null) {
+            throw new QueryParsingException(parseContext, "failed to find geo_point field [" + fieldName + "]");
+        }
+        if (!(fieldType instanceof GeoPointFieldMapper.GeoPointFieldType)) {
+            throw new QueryParsingException(parseContext, "field [" + fieldName + "] is not a geo_point field");
+        }
+        GeoPointFieldMapper.GeoPointFieldType geoFieldType = ((GeoPointFieldMapper.GeoPointFieldType) fieldType);
+
+        Query result;
+        if (Type.INDEXED.equals(type)) {
+            result = IndexedGeoBoundingBoxQuery.create(topLeft, bottomRight, geoFieldType);
+        } else if (Type.MEMORY.equals(type)) {
+            IndexGeoPointFieldData indexFieldData = parseContext.getForField(fieldType);
+            result = new InMemoryGeoBoundingBoxQuery(topLeft, bottomRight, indexFieldData);
+        } else {
+            throw new QueryParsingException(parseContext, "geo bounding box type [" + type
+                    + "] not supported, either 'indexed' or 'memory' are allowed");
+        }
+
+        if (queryName != null) {
+            parseContext.addNamedQuery(queryName, result);
+        }
+
+        return result;
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
 
-        builder.startObject(name);
-        builder.array(TOP_LEFT, box[LEFT], box[TOP]);
-        builder.array(BOTTOM_RIGHT, box[RIGHT], box[BOTTOM]);
+        builder.startObject(fieldName);
+        builder.array(GeoBoundingBoxQueryParser.TOP_LEFT, topLeft.getLon(), topLeft.getLat());
```

Comment:
Here we don't validate the points yet, but I guess this is okay since this will either pass the parser of just use this for logging after the refactoring.